File: kdb-wire/src/commonTest/kotlin/dev/kdb/wire/MalformedFrameTest.kt
package dev.kdb.wire

import kotlin.test.Test
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

/**
 * A frame's length prefix is attacker-controlled: it arrives whole over WebSocket (unlike the
 * stream reader, nothing has checked it against the buffer it came in) and can be truncated in a
 * capture file read by tooling. Every malformed shape has to end as a [WireDecodeException] the
 * caller can handle, not as an IndexOutOfBoundsException out of copyOfRange.
 *
 * Mirrors go/kdb/wire/malformed_frame_test.go - the two decoders read the same bytes and must
 * reject the same inputs.
 */
class MalformedFrameTest {
    private val codec = defaultWireCodec()

    /**
     * A syntactically valid 12-byte header in a buffer of [bufLen] bytes, whose length prefix
     * declares [declaredLen]. When the two disagree the frame is malformed exactly the way a
     * hostile or corrupt peer produces.
     */
    private fun framePrefix(
        bufLen: Int,
        declaredLen: Int,
        typeCode: Int = WireMessageType.HANDSHAKE.code.toInt(),
    ): ByteArray {
        val buf = ByteArray(bufLen)
        writeInt32Le(buf, 0, declaredLen)
        if (bufLen >= 12) {
            writeInt16Le(buf, 4, typeCode)
            writeInt16Le(buf, 6, KDB_WIRE_PROTOCOL_VERSION)
            writeInt32Le(buf, 8, 7)
        }
        return buf
    }

    private fun writeInt32Le(
        buf: ByteArray,
        offset: Int,
        value: Int,
    ) {
        buf[offset] = (value and 0xFF).toByte()
        buf[offset + 1] = ((value shr 8) and 0xFF).toByte()
        buf[offset + 2] = ((value shr 16) and 0xFF).toByte()
        buf[offset + 3] = ((value shr 24) and 0xFF).toByte()
    }

    private fun writeInt16Le(
        buf: ByteArray,
        offset: Int,
        value: Int,
    ) {
        buf[offset] = (value and 0xFF).toByte()
        buf[offset + 1] = ((value shr 8) and 0xFF).toByte()
    }

    @Test
    fun decodeRejectsDeclaredLengthLongerThanBuffer() {
        val ex = assertFailsWith<WireDecodeException> { codec.decode(framePrefix(20, 1000)) }
        assertTrue(
            ex.message!!.contains("declared length"),
            "unexpected message: ${ex.message}",
        )
    }

    @Test
    fun decodeHeaderRejectsDeclaredLengthLongerThanBuffer() {
        assertFailsWith<WireDecodeException> { codec.decodeHeader(framePrefix(20, 1000)) }
        // One byte short is still short: the boundary is exact, not approximate.
        assertFailsWith<WireDecodeException> { codec.decodeHeader(framePrefix(19, 20)) }
    }

    @Test
    fun decodeHeaderAcceptsExactAndOverlongBuffers() {
        val header = codec.decodeHeader(framePrefix(13, 13))
        kotlin.test.assertEquals(1, header.payloadLength)
        // A buffer longer than its declared length is legal for decodeHeader - the frame readers
        // hand over exact buffers, but nothing about the header itself is wrong.
        codec.decodeHeader(framePrefix(64, 13))
    }

    @Test
    fun decodeRejectsEmptyDeclaredPayload() {
        assertFailsWith<WireDecodeException> { codec.decode(framePrefix(64, 12)) }
    }

    @Test
    fun decodeHeaderRejectsUnknownMessageType() {
        assertFailsWith<WireDecodeException> {
            codec.decodeHeader(framePrefix(32, 32, typeCode = 0x7FFF))
        }
    }

    /**
     * The floor on a declared frame length is the header's own size. This was 8 while
     * FRAME_HEADER_SIZE has been 12, so lengths 8..11 - which Go has always rejected - were
     * accepted here, and then produced a negative payload length downstream.
     */
    @Test
    fun validateFrameLengthFloorIsTheHeaderSize() {
        for (tooSmall in listOf(0, 1, 7, 8, 9, 10, 11)) {
            assertFailsWith<FrameTooLargeException>("length $tooSmall was accepted") {
                validateFrameLength(tooSmall)
            }
        }
        validateFrameLength(12) // the smallest frame that can exist: header, no payload
        assertFailsWith<FrameTooLargeException> {
            validateFrameLength(DEFAULT_MAX_FRAME_BYTES + 1)
        }
    }

    @Test
    fun decodeHeaderRejectsNegativeDeclaredLength() {
        assertFailsWith<FrameTooLargeException> { codec.decodeHeader(framePrefix(32, -1)) }
    }
}

/**
 * The commit count in a CommitPush payload is peer-controlled and was handed straight to
 * ArrayList as a capacity, so four bytes could ask for a two-billion-element backing array.
 * Mirrors go/kdb/wire's TestDecodeCommitsDoesNotAllocateFromDeclaredCount.
 */
class CommitPushCodecMalformedTest {
    @Test
    fun decodeCommitsRejectsCountLargerThanPayloadCanHold() {
        // count = 0x7FFFFFFF with no commit bodies behind it
        val claimsTwoBillion = byteArrayOf(0xff.toByte(), 0xff.toByte(), 0xff.toByte(), 0x7f)
        assertFailsWith<IllegalArgumentException> {
            CommitPushCodec.decodeCommits(claimsTwoBillion)
        }
    }

    @Test
    fun decodeCommitsRejectsNegativeCount() {
        // A length prefix with the high bit set reads back as a negative Int, which used to take
        // out ArrayList's own argument check rather than producing a decode error.
        val negative = byteArrayOf(0xff.toByte(), 0xff.toByte(), 0xff.toByte(), 0xff.toByte())
        assertFailsWith<IllegalArgumentException> { CommitPushCodec.decodeCommits(negative) }
    }

    @Test
    fun decodeCommitsAcceptsPayloadTooShortForACount() {
        // Fewer than four bytes is "no commits", not an error - matching Go's DecodeCommits.
        kotlin.test.assertEquals(0, CommitPushCodec.decodeCommits(byteArrayOf(1, 2)).size)
    }

    @Test
    fun decodeCommitsAcceptsAnEmptyList() {
        val encoded = CommitPushCodec.encodeCommits(emptyList())
        kotlin.test.assertEquals(0, CommitPushCodec.decodeCommits(encoded).size)
    }
}

File: go/kdb/wire/malformed_frame_test.go
package wire_test

import (
	"encoding/binary"
	"strings"
	"testing"

	"github.com/limidus/kdb/go/kdb/wire"
)

// framePrefix builds a syntactically valid 12-byte header in a buffer of bufLen bytes, whose
// length prefix declares declaredLen. When declaredLen != bufLen the frame is malformed in
// exactly the way a hostile or corrupt peer produces.
func framePrefix(t *testing.T, bufLen, declaredLen int, msgType wire.MessageType) []byte {
	t.Helper()
	buf := make([]byte, bufLen)
	binary.LittleEndian.PutUint32(buf[0:], uint32(int32(declaredLen)))
	binary.LittleEndian.PutUint16(buf[4:], uint16(msgType))
	binary.LittleEndian.PutUint16(buf[6:], wire.KdbWireProtocolVersion)
	binary.LittleEndian.PutUint32(buf[8:], 7)
	return buf
}

// A frame whose prefix claims far more bytes than it carries must be a decode error, not a
// slice-bounds panic. Every caller derives the payload slice from PayloadLength, which is
// computed from the declared length, so this used to panic and - with no recover() anywhere on
// the connection-handling path - take the whole process down with it.
func TestDecodeRejectsDeclaredLengthLongerThanBuffer(t *testing.T) {
	c := wire.NewCodec(wire.EncodingKdbBinary)
	frame := framePrefix(t, 20, 1000, wire.MsgHandshake)

	msg, err := c.Decode(frame)
	if err == nil {
		t.Fatalf("expected a decode error, got message %#v", msg)
	}
	if !strings.Contains(err.Error(), "declared length") {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestDecodeHeaderRejectsDeclaredLengthLongerThanBuffer(t *testing.T) {
	if _, err := wire.DecodeHeader(framePrefix(t, 20, 1000, wire.MsgHandshake)); err == nil {
		t.Fatal("expected a decode error for a header declaring more bytes than the buffer holds")
	}
	// One byte short is still short: the boundary is exact, not approximate.
	if _, err := wire.DecodeHeader(framePrefix(t, 19, 20, wire.MsgHandshake)); err == nil {
		t.Fatal("expected a decode error for a buffer one byte shorter than declared")
	}
	// An exact-length frame with a real payload byte still decodes its header fine.
	exact := framePrefix(t, 13, 13, wire.MsgHandshake)
	h, err := wire.DecodeHeader(exact)
	if err != nil {
		t.Fatalf("exact-length frame: %v", err)
	}
	if h.PayloadLength != 1 {
		t.Fatalf("payloadLength: got %d, want 1", h.PayloadLength)
	}
	// A buffer longer than its declared length is legal for DecodeHeader - the frame readers
	// hand over exact buffers, but nothing about the header itself is wrong.
	if _, err := wire.DecodeHeader(framePrefix(t, 64, 13, wire.MsgHandshake)); err != nil {
		t.Fatalf("over-long buffer with a valid declared length: %v", err)
	}
}

// A frame declaring a zero-length payload while carrying trailing bytes gets past the
// "too short for payload" guard and would slice frame[13:12] - low above high, which panics
// exactly like an out-of-range high bound.
func TestDecodeRejectsEmptyDeclaredPayload(t *testing.T) {
	c := wire.NewCodec(wire.EncodingKdbBinary)
	if _, err := c.Decode(framePrefix(t, 64, wire.FrameHeaderSize, wire.MsgHandshake)); err == nil {
		t.Fatal("expected a decode error for a frame declaring an empty payload")
	}
}

// DecodePayloadEnvelope takes its header from the caller instead of re-deriving it, so it has
// to re-check the bound itself: a header that does not belong to the frame it is passed with
// must be an error rather than a panic. kdb-inspect pairs the two from separate calls.
func TestDecodePayloadEnvelopeRejectsMismatchedHeader(t *testing.T) {
	frame := framePrefix(t, 64, 64, wire.MsgHandshake)
	header, err := wire.DecodeHeader(frame)
	if err != nil {
		t.Fatal(err)
	}
	header.PayloadLength = 10_000 // as if it came from a different, larger frame
	if _, err := wire.DecodePayloadEnvelope(frame, header); err == nil {
		t.Fatal("expected an error for a header whose payload does not fit the frame")
	}
	header.PayloadLength = 0
	if _, err := wire.DecodePayloadEnvelope(frame, header); err == nil {
		t.Fatal("expected an error for a header declaring an empty payload")
	}
}

// The unknown-type and oversized rejections should stay ahead of any payload handling, so a
// garbage frame is refused on its header alone.
func TestDecodeHeaderRejectsUnknownMessageType(t *testing.T) {
	frame := framePrefix(t, 32, 32, wire.MessageType(0x7FFF))
	if _, err := wire.DecodeHeader(frame); err == nil {
		t.Fatal("expected an error for an unknown message type code")
	}
}

func TestDecodeHeaderRejectsNegativeAndOversizedLengths(t *testing.T) {
	for _, tc := range []struct {
		name     string
		declared int
	}{
		{"negative", -1},
		{"below header size", 4},
		{"above max frame bytes", wire.DefaultMaxFrameBytes + 1},
	} {
		t.Run(tc.name, func(t *testing.T) {
			if _, err := wire.DecodeHeader(framePrefix(t, 32, tc.declared, wire.MsgHandshake)); err == nil {
				t.Fatalf("expected an error for declared length %d", tc.declared)
			}
		})
	}
}
